packages/framework/immutable-merge/src/Merge.ts

```diff
@@ -51,76 +128,78 @@ export type IMergeRecursionHandler = (options: IMergeOptions, ...objs: any[]) =>
  * be `{}` - no recursion, `{ depth: -1 }` - recurse infinitely
  * @param objs - an array of objects to merge together
  */
-export function immutableMergeCore(options: IMergeOptions, ...objs: any[]): any {
-  const setToMerge = objs.filter(v => _isNonArrayObject(v, true));
-  const { depth = 0, recurse } = options;
-  const processSingle = options.processSingles && setToMerge.length === 1;
+function immutableMergeWorker(mergeOptions: RecursionOption | MergeOptions, singleMode: boolean, ...objs: any[]): any {
+  const setToMerge = objs.filter(v => v && getEntityType(v) === 'object' && Object.getOwnPropertyNames(v).length > 0);
+  const [options, mightRecurse] = normalizeOptions(mergeOptions);
+  const processSingle = singleMode && setToMerge.length === 1;
 
   // there is work to do if there is more than one object to merge or if we are processing single objects
-  if (setToMerge.length > 1 || processSingle) {
-    const nextOptions = depth || recurse ? { ...options, depth: depth ? depth - 1 : 0 } : undefined;
-    let hasChanged = false;
-
+  if (setToMerge.length > 1 || (processSingle && setToMerge.length === 1)) {
     // now assign everything to get the normal property precedence (and merge all the keys)
-    const result = Object.assign({}, ...setToMerge);
-
-    // if there is a possibility of needing the recurse, process the keys
-    for (const key in result) {
-      if (result.hasOwnProperty(key)) {
-        const originalVal = result[key];
-
-        // next options is only set if there is a possibility of recursion
-        if (nextOptions) {
-          const isObject = _isNonArrayObject(originalVal);
-          const recurseKey = recurse && recurse[key];
-          const customHandler = recurseKey && typeof recurseKey === 'function' ? recurseKey : false;
-
-          // if this key has a custom handler, or is an object eligible for recursion, recurse
-          if (customHandler || (isObject && (depth || recurseKey))) {
-            // collect all the keys, filtering out undefined values
-            const collectedObjects = setToMerge
-              .map((value: any) => {
-                return _collectVal(value && value[key], !!customHandler);
-              })
-              .filter(v => v !== undefined);
-
-            // either recurse to this function or call the custom handler if specified
-            const handler: IMergeRecursionHandler = customHandler || immutableMergeCore;
-            result[key] = handler(nextOptions, ...collectedObjects);
-            hasChanged = hasChanged || result[key] !== originalVal;
+    let result = processSingle ? undefined : Object.assign({}, ...setToMerge);
+    const processSet = result || setToMerge[0];
+
+    for (const key in processSet) {
+      if (processSet.hasOwnProperty(key)) {
```

Comment:
Similar concerns to getOwnPropertyNames. This probably works for all of our use cases I do wonder about people doing things directly via prototypal inheritance and whether the results end up being wrong.

Reply:
I believe this is the correct way to do it.  I just looked at the polyfill for Object.assign and it has the following:
```js
__assign = Object.assign || function(t) {
        for (var s, i = 1, n = arguments.length; i < n; i++) {
            s = arguments[i];
            for (var p in s) if (Object.prototype.hasOwnProperty.call(s, p))
                t[p] = s[p];
        }
        return t;
    };
```

TypeScript also has a lint error for unprotected key usages within in statements which is why I started adding these when doing the theming stuff.  Given the standard polyfill I think this is correct, it is simply that if your object can't be assigned (e.g. if it is a class) then it can't be merged.